src/socket.rs

```diff
@@ -20,16 +20,20 @@ pub trait Socket {
 
     fn send(&mut self, data: &[u8]) -> GDResult<()>;
     fn receive(&mut self, size: Option<usize>) -> GDResult<Vec<u8>>;
+
+    fn port(&self) -> u16;
 }
 
 pub struct TcpSocket {
     socket: net::TcpStream,
+    address: SocketAddr,
 }
 
 impl Socket for TcpSocket {
     fn new(address: &SocketAddr) -> GDResult<Self> {
         Ok(Self {
             socket: net::TcpStream::connect(address).map_err(|e| SocketConnect.context(e))?,
+            address: *address,
         })
```

Comment:
Would it be better to use `self.socket.peer_addr()` instead of implicitly copying the address? Maybe not as that is wrapped in a result.

Reply:
I don't think so, we already have the address and I guess (?) that internally `peer_addr` also does a copy.

Reply:
Yeah it most likely does.